HPCC-31213 Add plane name to MoveExternalFile, etc

wangkx opened the pull request:
The STD.File.MoveExternalFile moves files from one landing zone folder to anothor. The existing method can only specify a landing zone using host name. In this PR, a landing zone can be specified by plane name.
Similar functions are added to STD.File.CreateExternalDirectory and STD.File.DeleteExternalFile.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-31213 Add plane name to MoveExternalFile, etc

The STD.File.MoveExternalFile moves files from one
landing zone folder to anothor. The existing method
can only specify a landing zone using host name. In
this PR, a landing zone can be specified by plane
name. Also add plane name to DeleteExternalFile and
CreateExternalDirectory.

Revise based on review:
1. Change the checkExternalFileRights() to call getScopePermissions always.
2. Pass correct scopes to the checkExternalFileRights().
3. In the checkExternalFilePath(), if file->exists(), assume it's a
file and validate ignoring the tail.
4. If plane name is not specified, use the findDropZonePlane() and
useDropZoneRestriction.
5. In ecllibrary/std/File.ecl, make the planename optional for the
MoveExternalFile().
6. Add a comment for checkPlaneOrHost()
7. Fix smoke test failures: In the despray.ecl, remove the DeleteExternalFile
calls for malformed filenames and add a comment.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2747,21 +2751,60 @@ static void checkExternalFileRights(ICodeContext *ctx, CDfsLogicalFileName &lfn,
+        dlfn.setExternal(host,path); //for backward compatibility
+    dlfn.getExternalFilename(rfn);
+
+    if (checkScopePermissionIfDir)
```
see prev. comment. I think this should be done in implementMoveExternalFile, ensuring the lfn is always to the dir.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2747,21 +2751,60 @@ static void checkExternalFileRights(ICodeContext *ctx, CDfsLogicalFileName &lfn,
+    Owned<IPropertyTree> plane = checkPlaneOrHost(planename,location);
+    RemoteFilename fromrfn,torfn;
+    checkExternalFilePath(ctx,plane,location,frompath,true,true,true,fromrfn);
+    checkExternalFilePath(ctx,plane,location,topath,false,false,true,torfn);
```
see other comments, but if frompath is a directory, it implies topath should be too right?

wangkx replied:
Yes. And the access to the directory (not a parent) should be checked.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2774,14 +2817,25 @@ FILESERVICES_API void  FILESERVICES_CALL fsMoveExternalFile(ICodeContext * ctx,c
     CDfsLogicalFileName lfn;
     lfn.setExternal(location,path);
-    checkExternalFileRights(ctx,lfn,false,true);
+    checkExternalFileRights(ctx,lfn,false,false,true); //The path could be a directory. So, it may need to be checked as a scope. Will re-visit this when adding plane name here.
```
agree. CreateExternalDirectory also implies it must be a directory.
I think all cases need to ensure that the lfn passed to checking function is a directory.

jakesmith reviewed (commented):
@wangkx - please see comments.

wangkx commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2726,15 +2726,19 @@ FILESERVICES_API char *  FILESERVICES_CALL fsfResolveHostName(const char *hostna
 }
 
-static void checkExternalFileRights(ICodeContext *ctx, CDfsLogicalFileName &lfn, bool rd,bool wr)
+static void checkExternalFileRights(ICodeContext *ctx,CDfsLogicalFileName &lfn,bool checkScopePermissions,bool rd,bool wr)
```
Will do

wangkx commented:
@jakesmith I fixed one regress test failure in https://github.com/hpcc-systems/HPCC-Platform/pull/18275/commits/a20d0e7d3fe8b8957b27f8295a7a909914b590b3. That makes me wondering if I should do the similar changes for other exception related code I added. But, the old code also throws some exceptions. 
With the 2 new commits, there still exists a regress test failure when running despray.ecl. I think it is because I change the code from the 'rfn.setPath(ep,path)' to 'lfn.getExternalFilename(rfn);' for path '/var/lib/HPCCSystems/mydropzone/../W20240208-185950-persons'. Please advise.

wangkx commented:
Looks like the commit https://github.com/hpcc-systems/HPCC-Platform/pull/18275/commits/7834bcb0a35c8e40dc31fd74d88eca3f288d0fef fixes the regress test failures.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2726,70 +2726,152 @@ FILESERVICES_API char *  FILESERVICES_CALL fsfResolveHostName(const char *hostna
+        RemoteFilename rfn;
+        lfn.getExternalFilename(rfn);
+        Owned<IFile> file = createIFile(rfn);
+        if (!file->exists())
```
strictly, I think this is a security leak. It means unauthorized users can discover which files do/don't exist.

It should not leak this information for unauthorized users.
To overcome this, I think you need to assume it's a file and validate the parent scope 1st.
Then check if directory, and only then check the tail if it is.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2726,70 +2726,152 @@ FILESERVICES_API char *  FILESERVICES_CALL fsfResolveHostName(const char *hostna
+        Owned<IPropertyTree> plane = checkPlaneOrHost(planename,location);
+        RemoteFilename fromrfn,torfn;
+        checkExternalFilePath(ctx,plane,location,frompath,true,true,fromrfn);
+        checkExternalFilePath(ctx,plane,location,topath,false,true,torfn);
```
haven't thought about too deeply, but feels like dropzone vs paths within dropzones checking + perm. checking functionality should already be covered by existing implemented functionality e.g. validateDZPathScopePermsAndLegacyPhysicalPerms etc.

Plus, see other comment re. if if unknown whether file or dir, assume file, and check user has perms to access parent dir. 1st, before returning checking returning error that file doesn't exist.

jakesmith replied:
> validateDZPathScopePermsAndLegacyPhysicalPerms

@wangkx I realize methods like validateDZPathScopePermsAndLegacyPhysicalPerms are currently in SMCLib and use IEspContext's and therefore inaccessible to the fileservices plugin.

But if this is a common functional need, that currently requires very similar functionality to be duplicated, it suggests it should be commoned up somewhere, where it is accessible by both.

It may be that it is best for this PR to leave it a duplicated functionality, but open a JIRA and revisit to refactor to avoid the duplication/future confusion.

wangkx replied:
I opened https://track.hpccsystems.com/browse/HPCC-31280 to refactor to avoid the duplication/future confusion.

jakesmith reviewed (commented):
@wangkx - returning to you for now. 
I think approach should be changed with regard to validating a path that may be a dir..
And, want to check this is not duplicating existing functionality like validateDZPathScopePermsAndLegacyPhysicalPerms.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2848,19 +2841,11 @@ FILESERVICES_API void  FILESERVICES_CALL fsDeleteExternalFile(ICodeContext * ctx
             throw makeStringExceptionV(-1,"Cannot resolve location %s",location);
         CDfsLogicalFileName lfn;
         lfn.setExternal(location,path);
+        checkExternalFileRights(ctx,lfn.get(),false,true);
```
re. this previous comment : https://github.com/hpcc-systems/HPCC-Platform/pull/18275#discussion_r1485986966

It looks like it now, passes the full filename here (e.g. scope1::scope2::filename), then scope checks it, including "filename" tail.

The previous comment was suggesting, that since you do not know whether the tail is a directory or a filename, you should assume it's a filename and scope check the parent ("scope1::scope2"), only then once you have proved the user has access to that scope, should you check if the tail ("filename" in this case) is a directory, if it is, you need to scope check that too, if not, you are done.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2770,17 +2770,10 @@ static void checkExternalFilePath(ICodeContext *ctx,IPropertyTree *plane,const c
+    //If the path is a folder, check if it can be accessed.
+    //If the path is a file, the LDAP will skip the file layer (because it is undefined in the scope access setting)
+    //and check if its scope can be accessed.
+    checkExternalFileRights(ctx,dlfn.get(),rd,wr);
```
see comment below (related to previous comment : https://github.com/hpcc-systems/HPCC-Platform/pull/18275#discussion_r1485986966)

If this function is to be used in context where you do not know whether the user supplied path is a path to a directory, or a path to a file, you need to validate access to the parent scope e.g. "scope1::scope2" from "scope1::scope2::tail", then check what 'tail' is, and then only if it is a directory, scope check it.

wangkx commented:
@jakesmith I squashed the old commits into [1f7152e](https://github.com/hpcc-systems/HPCC-Platform/pull/18275/commits/1f7152ef5e44053fe2a4ce0552fb743792e95cea). In [ad1b09e](https://github.com/hpcc-systems/HPCC-Platform/pull/18275/commits/ad1b09e448f0b3a4684685dc5df46519b0d196ff), I changed the code to pass different scopes to checkExternalFileRights(). Since the code still throws the exception if a file is not found and the user can access the scope, a smote test failure is occurred as expected. The [7b3114f](https://github.com/hpcc-systems/HPCC-Platform/pull/18275/commits/7b3114f2c40387c5ce22b34e4dec21c178a566ae) is added to fix the failure. Please review.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2726,42 +2726,91 @@ FILESERVICES_API char *  FILESERVICES_CALL fsfResolveHostName(const char *hostna
+    }
+}
+
+static IPropertyTree *checkPlaneOrHost(const char *planeName,const char *host)
```
see comment where used, I think in next refactoring, this new method should go.

jakesmith commented on ecllibrary/std/File.ecl:
```diff
@@ -358,9 +358,10 @@ EXPORT RfsAction(varstring server, varstring query) :=
-EXPORT MoveExternalFile(varstring location, varstring frompath, varstring topath) :=
-    lib_fileservices.FileServices.MoveExternalFile(location, frompath, topath);
+EXPORT MoveExternalFile(varstring location, varstring frompath, varstring topath, varstring planename) :=
+    lib_fileservices.FileServices.MoveExternalFile(location, frompath, topath, planename);
```
making it non-optional is going to break existing ECL.
When these (plane names) were introduced to the spray calls they were added as optional parameters for that reason.

jakesmith reviewed (changes requested):
@wangkx - please see comments.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2774,19 +2829,44 @@ FILESERVICES_API void  FILESERVICES_CALL fsMoveExternalFile(ICodeContext * ctx,c
+        checkExternalFileRights(ctx,lfn.get(),false,true);
+        s.appendf("fsDeleteExternalFile: File %s not found.')",path);
+        WUmessage(ctx,SeverityInformation,nullptr,s.str());
+        return;
```
not clear why this is coded differently to MoveExternalFile.
This is check file exists 1st, and then checking the full scope (including the supposed filename tail).

Afaics, it should be coded in the same way that moveExternalFile is, i.e. only check full scope if (f->exists() && (f->isDirectory()==fileBool::foundYes)).

In fact, why not use checkExternalFilePath (with plane=nullptr) ?

jakesmith replied:
*But* - this JIRA/PR is titled "Add plane parameter to STD.File.MoveExternalFile"

You should either:
1) not make changes to DeleteExternalFile in this PR, and make it a separate JIRA/PR, OR
2) Change title/comments, and add a plane parameter to DeleteExternalFile at same time.

jakesmith replied:
> @jakesmith in the existing commits, both DeleteExternalFile and CreateExternalDirectory are changed because the changes in the checkExternalFileRights(). I do not want the changes in the checkExternalFileRights() to break those 2 functions. The original intention is not adding a plane parameter in those functions in this PR. If you want, I can roll back the changes in DeleteExternalFile and CreateExternalDirectory. Please let me know.

@wangkx - Given the knock on impact change to MoveExternalFile has had, and the required changes to DeleteExternalFile and CreateExternalDirectory, and the desire to add plane to those anyway, I would make them all consistent and add a plane parameter to both of these in this PR (and retitle the JIRA and commit.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2726,42 +2726,97 @@ FILESERVICES_API char *  FILESERVICES_CALL fsfResolveHostName(const char *hostna
+}
+
+static IPropertyTree *checkPlaneOrHost(const char *planeName,const char *host,const char *path)
+{
```
minor: can you add a // TBD comment here that this function is very similar to getDropZoneAndValidateHostAndPath in TpCommon, and both should be replaced/moved so common?

jakesmith reviewed (changes requested):
@wangkx - pl ease see latest comments.

wangkx commented:
@jakesmith looks like that the smoke test failure is because of the code:
```
static void checkExternalFilePath(ICodeContext *ctx,IPropertyTree *plane,const char *host,
    const char *path,bool rd,bool wr,RemoteFilename &rfn)
{
    if (containsRelPaths(path)) //Detect a path like: a/../../../f
        throw makeStringExceptionV(-1,"Invalid file path %s",path);
```
please advise. Should I add a try/catch to switch the exception to a log.

wangkx commented:
The smoke test failure 'Invalid file path' is fixed by [8a70be3](https://github.com/hpcc-systems/HPCC-Platform/pull/18275/commits/8a70be390baee3d8851e94215c8d1a1db85af212). But, the code got another smoke test failure (see below) which may happen if the smoke test runs on a cloud env or a bare metal env with isDropZoneRestrictionEnabled. 

Error:
    188. Test: despray.ecl
    '--- despray.xml
    +++ despray.xml
    @@ -1,3 +1,4 @@
    +<Exception><Code>4294967295</Code><Source>Roxie</Source><Message>DropZone Plane not found for host . path /var/lib/HPCCSystems/mydropzona/W20240217-030601-1-persons.</Message></Exception>

jakesmith commented:
> @jakesmith looks like that the smoke test failure is because of the code:
> 
> ```
> static void checkExternalFilePath(ICodeContext *ctx,IPropertyTree *plane,const char *host,
>     const char *path,bool rd,bool wr,RemoteFilename &rfn)
> {
>     if (containsRelPaths(path)) //Detect a path like: a/../../../f
>         throw makeStringExceptionV(-1,"Invalid file path %s",path);
> ```
> 
> please advise. Should I add a try/catch to switch the exception to a log?


@wangkx - I see the last commit has moved the relative check from within checkExternalFilePath to callers....
What exactly failed (and why) ?

jakesmith reviewed (commented):
@wangkx - see comments/questions

wangkx commented:
> I see the last commit has moved the relative check from within checkExternalFilePath to callers....
What exactly failed (and why) ?

It happened when running the following code in despray.ecl. Because the DestFile4 contains the 'mydropzone/./', the code before that commit throws: <Exception><Code>4294967295</Code><Source>Roxie</Source><Message>Invalid file path /var/lib/HPCCSystems/mydropzone/./W20240216-193252-5-persons</Message></Exception>
 
// This should fail based on '/./' used in target path
DestFile4 := dropzonePath + './' + DestFileName;
dst4 := NOFOLD(DATASET([{SourceFile, DestFile4, SrcAddrIp, True, '', ''}], rec));

p4 := PROJECT(NOFOLD(dst4), t(LEFT));

c4 := CATCH(NOFOLD(p4), ONFAIL(TRANSFORM(rec,
                                 SELF.result := 'Fail',
                                 SELF.destFile := DestFile4,
                                 SELF.sourceFile := SourceFile,
                                 SELF.ip := SrcAddrIp,
                                 SELF.allowOverwrite := True,
                                 SELF.msg := FAILMESSAGE
                                )));
#if (VERBOSE = 1)
    o4 := output(c4);
#else
    o4 := output(c4, {result});
#end

wangkx commented:
> That's because despray.ecl is trying to delete files for the invalid tests (i.e. those that are specificially outside of a dropzone). See DestFile6.The test is wrong, these files will never have been created. The test(s) should be fixed.

I think that those tests were added intentionally (including the tests for DestFile4 and DestFile6). I am not sure which tests should be fixed.

wangkx commented:
I guess I should remove the following code from the despray.ecl.

/// FileServices.DeleteExternalFile('.', DestFile4),
  ///  FileServices.DeleteExternalFile('.', DestFile5),
  ///  FileServices.DeleteExternalFile('.', DestFile6),
  ///  FileServices.DeleteExternalFile('.', DestFile7),
  ///  FileServices.DeleteExternalFile('.', DestFile8),

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -2750,9 +2750,6 @@ static void checkExternalFileRights(ICodeContext *ctx,const char *scope,bool rd,
 {
-    if (containsRelPaths(path)) //Detect a path like: a/../../../f
-        throw makeStringExceptionV(-1,"Invalid file path %s",path);
-
```
@wangkx - Now the rogue DeleteExternalFile's are removed (commented out), that was the root cause of the regression suite failure, is this movement of this relative check correct? 
If not should move back.
If it is, when does it cause failures?

wangkx replied:
The relative check has been moved back to here.

jakesmith reviewed (approved):
@wangkx - looks fine. Please squash.

wangkx commented:
The commits are squashed.
@ghalliday I think the windows build failure is not related to this PR. Please check and merge. 
This is a dependent PR for https://github.com/hpcc-systems/HPCC-Platform/pull/18292,

ghalliday commented:
@wangkx please can you document the changes in the jira

wangkx commented:
@ghalliday the changes are documented inside https://track.hpccsystems.com/browse/HPCC-31213. If I miss anything, please let me know.
